python/tk_3de4/menu_generation.py

```diff
@@ -0,0 +1,216 @@
+# Copyright (c) 2015 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+Menu handling for 3DE4
+
+"""
+from collections import defaultdict
+from itertools import count
+import os
+import sys
+import unicodedata
+
+import sgtk
+from sgtk.platform.qt import QtGui, QtCore
+
+
+class MenuGenerator(object):
+    """
+    Menu generation functionality for 3DE4
+    """
+    root_ui_item = "Main Window::Shotgun"
+
+    def __init__(self, engine):
+        """
+        Initialise the class.
+
+        :param engine: The shotgun engine instance.
+        """
+        self._engine = engine
+        self._current_menu_index = count()
+        self.custom_scripts_dir_path = os.environ["TK_3DE4_MENU_DIR"]
+
+    ##########################################################################################
+    # public methods
+
+    @property
+    def logger(self):
+        """
+        Get the logger instance.
+
+        :returns: The logger associated to the engine.
+        """
+        return self._engine.logger
+
+    def create_menu(self):
+        """
+        Render the entire Shotgun menu.
+        In order to have commands enable/disable themselves based on the enable_callback,
+        re-create the menu items every time.
+        """
+        self.logger.info("Creating Shotgun menu...")
+
+        # Get temp folder path and create it if needed.
+        if not os.path.isdir(self.custom_scripts_dir_path):
+            self.logger.debug("Creating menu directory %s", self.custom_scripts_dir_path)
+            os.makedirs(self.custom_scripts_dir_path)
+        # Clear it.
+        for item in os.listdir(self.custom_scripts_dir_path):
+            self.logger.debug("Removing menu file: %s", item)
+            os.remove(os.path.join(self.custom_scripts_dir_path, item))
+
+        menu_items = []
+        for cmd_name, cmd_details in self._engine.commands.items():
+             menu_items.append(AppCommand(cmd_name, cmd_details))
```

Comment:
Maybe switch to list comprehension

```suggestion
        menu_items = [
            AppCommand(cmd_name, cmd_details)
            for cmd_name, cmd_details in self._engine.commands.items()
        ]
```